### PR TITLE
Warnings as errors in akka-actor

### PR DESCRIPTION
**Pull request by raboof**
Found an actual (though obscure) bug in ByteIterator \o/

Refs #26088

### CODE REVIEW

File: akka-actor/src/main/scala/akka/dispatch/Dispatcher.scala [hunk 1]
```diff
@@ -41,6 +44,10 @@ class Dispatcher(
     def copy(): LazyExecutorServiceDelegate = new LazyExecutorServiceDelegate(factory)
   }
 
+  /**
+   * Updated via esUpdater
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/dispatch/Dispatcher.scala]: ?

File: akka-actor/src/main/scala-2.12/akka/util/ByteIterator.scala [hunk 2]
```diff
@@ -460,7 +460,11 @@ abstract class ByteIterator extends BufferedIterator[Byte] {
 
   override def indexWhere(p: Byte => Boolean): Int = indexWhere(p, 0)
   override def indexWhere(p: Byte => Boolean, from: Int): Int = {
-    var index = from
+    var index = 0
+    while (index < from) {
+      next()
+      index += 1
+    }
```
Comment (johanandren) [inline akka-actor/src/main/scala-2.12/akka/util/ByteIterator.scala]: ups, 👍

File: akka-actor/src/main/scala/akka/routing/SmallestMailbox.scala [hunk 3]
```diff
@@ -77,6 +79,8 @@ class SmallestMailboxRoutingLogic extends RoutingLogic {
     }
   }
 
+  // TODO should we rewrite this not to use isTerminated?
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/routing/SmallestMailbox.scala]: given that the method itself is `isTerminated` maybe not?

**Review by johanandren — APPROVED** [review]

File: akka-actor/src/main/scala/akka/dispatch/Dispatcher.scala (hunk 1 above)
Comment (raboof) [reply akka-actor/src/main/scala/akka/dispatch/Dispatcher.scala]: At first glance this `var` would not seem to be updated anywhere - but in fact it is, via the esUpdater `AtomicReferenceFieldUpdater` below - I'll make the comment a bit clearer :)

Comment (johanandren) [reply akka-actor/src/main/scala/akka/dispatch/Dispatcher.scala]: Ah, I thought it was a comment added by a tool of some kind

File: akka-actor/src/main/scala-2.12/akka/util/ByteIterator.scala [hunk 4]
```diff
@@ -460,7 +460,11 @@ abstract class ByteIterator extends BufferedIterator[Byte] {
 
   override def indexWhere(p: Byte => Boolean): Int = indexWhere(p, 0)
   override def indexWhere(p: Byte => Boolean, from: Int): Int = {
-    var index = from
+    var index = 0
```
Comment (patriknw) [inline akka-actor/src/main/scala-2.12/akka/util/ByteIterator.scala]: create a separate issue for this bug, for reference in release notes

File: akka-actor/src/main/scala/akka/actor/ActorRefProvider.scala [hunk 5]
```diff
@@ -532,7 +532,7 @@ private[akka] class LocalActorRefProvider private[akka] (
     override def stop(): Unit = {
       causeOfTermination.trySuccess(
         Terminated(provider.rootGuardian)(existenceConfirmed = true, addressTerminated = true)) //Idempotent
-      terminationPromise.tryCompleteWith(causeOfTermination.future) // Signal termination downstream, idempotent
+      terminationPromise.completeWith(causeOfTermination.future) // Signal termination downstream, idempotent
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/ActorRefProvider.scala]: why is this change needed? can we be sure that it hasn't already been completed?

File: akka-actor/src/main/scala/akka/actor/dungeon/ChildrenContainer.scala [hunk 6]
```diff
@@ -140,7 +140,7 @@ private[akka] object ChildrenContainer {
     }
 
     override def toString =
-      if (c.size > 20) c.size + " children"
+      if (c.size > 20) c.size.toString + " children"
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/dungeon/ChildrenContainer.scala]: Will this result in unnecessary boxing or wrapping (allocating objects) if it's a primitive `int`? In Java one would use `String.valueOf(i)`. I't probably not important here but might be something to watch out for in any hot path (and there we probably already had that problem unless Scala did something very clever with `int + string`.

File: akka-actor/src/main/scala/akka/routing/ScatterGatherFirstCompleted.scala [hunk 7]
```diff
@@ -54,9 +54,9 @@ private[akka] final case class ScatterGatherFirstCompletedRoutees(
       val promise = Promise[Any]()
       routees.foreach {
         case ActorRefRoutee(ref) =>
-          promise.tryCompleteWith(ref.ask(message))
+          promise.completeWith(ref.ask(message))
         case ActorSelectionRoutee(sel) =>
-          promise.tryCompleteWith(sel.ask(message))
+          promise.completeWith(sel.ask(message))
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/routing/ScatterGatherFirstCompleted.scala]: is this correct? I think it sends out many requests and whatever is completed first will complete the promise, but other replies may arrive later

**Review by patriknw — COMMENTED** [review]
Looking good, but perhaps let's wait with merging this until after the release to be on the safe side. Possibly creating a separate PR fixing the bug, if we think it's important to include in the release.

File: akka-actor/src/main/scala/akka/actor/ActorRefProvider.scala (hunk 5 above)
Comment (raboof) [reply akka-actor/src/main/scala/akka/actor/ActorRefProvider.scala]: `tryCompleteWith` is deprecated in 2.13 because apparently `completeWith` always had these semantics already (unlike success/failure/complete): tested with `sbt ++2.11.12 console`:

```
import scala.concurrent._
p.completeWith(Future.successful(37))
p.completeWith(Future.successful(42))
implicit val ec = ExecutionContext.global
p.future.foreach(println)
37
```

Comment (patriknw) [reply akka-actor/src/main/scala/akka/actor/ActorRefProvider.scala]: ah, my misunderstanding then

Comment (raboof) [reply akka-actor/src/main/scala/akka/actor/ActorRefProvider.scala]: yeah, very confusing, surprised me as well

**Review by helena — APPROVED** [review]
LGTM

**Comment by raboof** [comment]
(reminder: let's merge after releaseing 2.5.22)
